File: sdks/javascript/examples/stateful-actor.ts
import { Actor, Message, MessageType, Capability } from '../src';

interface CounterState {
    count: number;
}

class StatefulActor extends Actor {
    static get name() { return 'counter'; }

    async onStart(): Promise<void> {
        this.require(Capability.STATE_READ, Capability.STATE_WRITE);
        const existing = await this.state.getJSON<CounterState>('counter');
        if (!existing) {
            await this.state.setJSON<CounterState>('counter', { count: 0 });
        }
    }

    async handle(sender: string, message: Message): Promise<Message | void> {
        if (message.type !== MessageType.CUSTOM) return;

        const state = await this.state.getJSON<CounterState>('counter') || { count: 0 };

        switch (message.payload.action) {
            case 'increment':
                state.count++;
                await this.state.setJSON('counter', state);
                return Message.custom({ count: state.count });

            case 'decrement':
                state.count--;
                await this.state.setJSON('counter', state);
                return Message.custom({ count: state.count });

            case 'get':
                return Message.custom({ count: state.count });

            default:
                return Message.custom({ error: 'Unknown action' });
        }
    }
}

const actor = new StatefulActor({ name: 'counter' });
actor.start().catch(console.error);

File: sdks/python/examples/stateful_actor.py
from aether_sdk import Actor, Message, MessageType, Capability


class StatefulActor(Actor):
    @classmethod
    def name(cls) -> str:
        return "stateful"
    
    def __init__(self):
        super().__init__()
        self.require(Capability.STATE_READ, Capability.STATE_WRITE)
    
    async def on_start(self) -> None:
        count = await self.state.get_json("counter")
        if count is None:
            await self.state.set_json("counter", 0)
    
    async def handle_message(self, sender: str, message: Message) -> Message:
        if message.type == MessageType.CUSTOM:
            action = message.payload.get("action")
            
            if action == "increment":
                count = await self.state.get_json("counter") or 0
                count += 1
                await self.state.set_json("counter", count)
                
                return Message(
                    type=MessageType.CUSTOM,
                    payload={"counter": count}
                )
            
            elif action == "get":
                count = await self.state.get_json("counter") or 0
                
                return Message(
                    type=MessageType.CUSTOM,
                    payload={"counter": count}
                )
            
            elif action == "reset":
                await self.state.set_json("counter", 0)
                
                return Message(
                    type=MessageType.CUSTOM,
                    payload={"counter": 0}
                )
        
        return None


if __name__ == "__main__":
    import asyncio
    
    async def main():
        actor = StatefulActor()
        await actor.on_start()
        
        for i in range(3):
            msg = Message(
                type=MessageType.CUSTOM,
                payload={"action": "increment"}
            )
            response = await actor.handle_message("test", msg)
            print(f"After increment {i+1}: counter = {response.payload['counter']}")
        
        msg = Message(
            type=MessageType.CUSTOM,
            payload={"action": "get"}
        )
        response = await actor.handle_message("test", msg)
        print(f"Final counter value: {response.payload['counter']}")
    
    asyncio.run(main())
